### PR TITLE
URI Parser v2

### PR DESCRIPTION
**Pull request by nyamsprod**
- Remove everything which is deprecated
- Symplify URI parsing using RFC3986 regular expression
- Make the Parser final
- Make protected methods private
- small bug fix within the test suite following the switch to RFC3986 regular expression

Question should we bump PHP version requirement to PHP7.1 ?

First benchmark show another performance boots to be validated with blackfire.io @kelunik or @trowski

### CODE REVIEW

File: src/Parser.php [hunk 1]
```diff
@@ -433,263 +417,18 @@ protected function filterHost($host)
      *
      * @return null|int
      */
-    protected function filterPort($port)
+    private function filterPort($port)
     {
         static $pattern = '/^[0-9]+$/';
```
Comment (kelunik) [inline src/Parser.php]: Leftover.

**Comment by kelunik** [comment]
Not sure whether requiring PHP 7.1 is warranted. What's the benefit except for one private constant over it being public?

File: src/Parser.php (hunk 1 above)
Comment (nyamsprod) [reply src/Parser.php]: good catch

File: benchmark/parse.php [hunk 2]
```diff
@@ -1,6 +1,9 @@
 <?php
 
-require __DIR__.'/../vendor/autoload.php';
+require __DIR__.'/../src/Exception/ParserException.php';
+require __DIR__.'/../src/Exception/MissingIdnSupport.php';
+require __DIR__.'/../src/Parser/UriParser.php';
+require __DIR__.'/../src/functions.php';
```
Comment (kelunik) [inline benchmark/parse.php]: Why?

Comment (nyamsprod) [reply benchmark/parse.php]: Because when using webgrind the composer stuffs get in the way of the result.

**Comment by nyamsprod** [comment]
@kelunik I'll be merging and bumping to the new version next week unless there's still some blockers here :thinking: . And also PHP7.0 support is dropped since last time we spoke it went in EOL mode :smile:

File: .scrutinizer.yml [hunk 3]
```diff
@@ -1,17 +1,25 @@
...
-        runs: 2
+        runs: 3
     php_code_coverage: false
     php_loc:
         enabled: true
         excluded_dirs: [tests, vendor]
     php_cpd:
         enabled: true
-        excluded_dirs: [tests, vendor]
+        excluded_dirs: [tests, vendor]
```
Comment (tgalopin) [inline .scrutinizer.yml]: Missing end line

**Comment by kelunik** [comment]
I won't have time to review, sorry. But sounds good. 👍

File: README.md [hunk 4]
```diff
@@ -4,17 +4,40 @@ League Uri Parser
 [![Build Status](https://img.shields.io/travis/thephpleague/uri-parser/master.svg?style=flat-square)](https://travis-ci.org/thephpleague/uri-parser)
 [![Latest Version](https://img.shields.io/github/release/thephpleague/uri-parser.svg?style=flat-square)](https://github.com/thephpleague/uri-parser/releases)
 
-This package contains a userland PHP uri parser compliant with [RFC 3986](http://tools.ietf.org/html/rfc3986).
+This package contains a userland PHP uri parser and builder compliant with:
```
Comment (tgalopin) [inline README.md]: URI?

File: README.md [hunk 5]
```diff
@@ -4,17 +4,40 @@ League Uri Parser
...
 -------
 
 You need:
 
-- **PHP >= 7.0** but the latest stable version of PHP is recommended
+- **PHP >= 7.1.3** but the latest stable version of PHP is recommended
 
-
-While the library no longer requires the `ext/intl` extension, it is strongly advise to install this extension if you are dealing with URIs containing non-ASCII host. Without the extension, the parser will throw an exception if such URI is parsed.
+While the library no longer requires the `ext/intl` extension, it is strongly advise to install this extension if you are dealing with URIs containing non-ASCII host. Without the extension, the parser will throw a `InvalidURI` exception when trying to parse such URI.
```
Comment (tgalopin) [inline README.md]: advised?

File: README.md [hunk 6]
```diff
@@ -27,12 +50,198 @@ $ composer require league/uri-parser
 Documentation
 ---------
 
-Full documentation can be found at [uri.thephpleague.com](http://uri.thephpleague.com).
+### URI Parsing
+
+~~~php
```
Comment (tgalopin) [inline README.md]: Is this a typo (to not use backtick)?

File: README.md (hunk 4 above)
Comment (nyamsprod) [reply README.md]: yes URI because RFC3986 talks about URI and URLs are bit a subset of URIs.

File: README.md (hunk 6 above)
Comment (nyamsprod) [reply README.md]: not a typo both are supported by Jenkyll but I'll normalized

File: README.md (hunk 4 above)
Comment (tgalopin) [reply README.md]: I mean, in uppercase :) ?

File: .gitignore [hunk 7]
```diff
@@ -3,3 +3,5 @@ build
 docs
 composer.lock
 .idea
+.php_cs.cache
+.phpunit.result.cache
```
Comment (tgalopin) [inline .gitignore]: Missing end line

File: README.md (hunk 4 above)
Comment (nyamsprod) [reply README.md]: ah yes ... let me fix that :smile:

File: benchmark/parse.php [hunk 8]
```diff
@@ -1,19 +1,17 @@
...
  * file that was distributed with this source code.
  */
-require __DIR__.'/../vendor/autoload.php';
+
+require __DIR__.'/../src/Parser/RFC3986.php';
 
 $uri = 'https://uri.thephpleague.com/5.0';
 for ($i = 0; $i < 100000; $i++) {
-    League\Uri\parse($uri);
+    RFC3986::parse($uri);
```
Comment (tgalopin) [inline benchmark/parse.php]: The namespace is missing.

File: benchmark/build.php [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+require __DIR__.'/../src/Parser/RFC3986.php';
+
+$components = [
+    'scheme' => 'http',
+    'host' => 'uri.thephpleague.com',
+    'path' => '/5.0',
+];
+
+for ($i = 0; $i < 100000; $i++) {
+    RFC3986::build($components);
```
Comment (tgalopin) [inline benchmark/build.php]: The namespace is missing.

**Review by tgalopin — CHANGES_REQUESTED** [review]

**Comment by tgalopin** [comment]
There is indeed a clear performance boost:

Before: https://blackfire.io/profiles/54c03b2b-46c8-4d9b-a6d1-7b3ffd4d5b13/graph

After: https://blackfire.io/profiles/ce6edc6d-d7f5-48cb-8225-5026f7738cd2/graph

File: benchmark/build.php (hunk 9 above)
Comment (nyamsprod) [reply benchmark/build.php]: resolve in the next commit. good catch

File: benchmark/parse.php (hunk 8 above)
Comment (nyamsprod) [reply benchmark/parse.php]: good catch

**Comment by nyamsprod** [comment]
@tgalopin the namespace are organized to simplify URI packages structure.

So the same structure is/will be used for

- URI Components v2 https://github.com/thephpleague/uri-components/tree/master/src
- URI Schemes v2 https://github.com/thephpleague/uri-schemes/tree/develop/src

is already used for

- URI Query parser v1 <- https://github.com/thephpleague/uri-query-parser/

All the other packages are being deprecated/phase out (3 packages will be removed)

```php
<?php
use League\Uri\Component\Query;
use League\Uri\Parser\RFC3986;
use League\Uri\Parser\QueryString;
use League\Uri\Resolver;
use League\Uri\Uri;
```

- The base namespace is reserved for full URI objects, interfaces and Helper classes.
- URI Components have their own namespace.
- URI parsers have their own namespace too regardless of their package.

The goal is to release all remaining v2 package before june... if I can :smile: . Most of the coding is done only reviewing and align packages remains to be done.

**Pull request merged**